wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Transactions.hs

```diff
@@ -0,0 +1,162 @@
+
+castSortingDirection :: S.SortOperation ix a -> TxMeta.SortDirection
+castSortingDirection (S.SortByIndex srt _) = case srt of
+    S.SortAscending  -> TxMeta.Ascending
+    S.SortDescending -> TxMeta.Descending
+
+castFiltering :: Maybe (F.FilterOperation ix V1.Transaction) -> TxMeta.FilterOperation ix
+castFiltering mfop = case mfop of
+    Nothing -> TxMeta.NoFilterOp
+    Just fop -> case fop of
+        (F.FilterByIndex q)         -> TxMeta.FilterByIndex q
+        (F.FilterByPredicate prd q) -> TxMeta.FilterByPredicate (castFilterOrd prd) q
+        (F.FilterByRange q w)       -> TxMeta.FilterByRange q w
+        (F.FilterIn ls)             -> TxMeta.FilterIn ls
+
+castFilterOrd :: F.FilterOrdering -> TxMeta.FilterOrdering
+castFilterOrd pr = case pr of
+    F.Equal            -> TxMeta.Equal
+    F.GreaterThan      -> TxMeta.GreaterThan
+    F.GreaterThanEqual -> TxMeta.GreaterThanEqual
+    F.LesserThan       -> TxMeta.LesserThan
+    F.LesserThanEqual  -> TxMeta.LesserThanEqual
+
+metaToTx :: Kernel.DB -> Word -> Word -> TxMeta -> V1.Transaction
+metaToTx db k current TxMeta{..} =
+    V1.Transaction {
+        txId = V1 _txMetaId,
+        txConfirmations = confirmations,
+        txAmount = V1 _txMetaAmount,
+        txInputs = toPayDistr <$> _txMetaInputs,
+        txOutputs = toPayDistr <$> _txMetaOutputs,
+        txType = if _txMetaIsLocal then V1.LocalTransaction else V1.ForeignTransaction,
+        txDirection = if _txMetaIsOutgoing then V1.OutgoingTransaction else V1.IncomingTransaction,
+        txCreationTime = V1 _txMetaCreationAt,
+        txStatus = status
+    }
+
+        where
+            hdAccountId = HD.HdAccountId (HD.HdRootId $ InDb _txMetaWalletId)
+                                        (HD.HdAccountIx _txMetaAccountId)
+
+            toPayDistr :: (Address, Coin) -> V1.PaymentDistribution
+            toPayDistr (addr, c) = V1.PaymentDistribution (V1 addr) (V1 c)
+
+            mSlot = Kernel.accountTxSlot db hdAccountId _txMetaId
+            isPending = Kernel.accountIsTxPending db hdAccountId _txMetaId
+
+            (status, confirmations) = buildDynamicTxMeta mSlot k current isPending
+
+buildDynamicTxMeta :: Maybe SlotId -> Word -> Word -> Bool -> (V1.TransactionStatus, Word)
+buildDynamicTxMeta mSlot k currentSlot isPending = case isPending of
+    True  -> (V1.Applying, 0)
+    False ->
+        case mSlot of
+        Nothing     -> (V1.WontApply, 0)
+        Just (SlotId (EpochIndex w64) (UnsafeLocalSlotIndex w16)) ->
+            case ((fromIntegral currentSlot) - w64*(fromIntegral k) + (fromIntegral w16) >= fromIntegral k) of -- TODO: fix
+            True  -> (V1.InNewestBlocks, fromIntegral w64) -- TODO: fix
+            False -> (V1.Persisted, fromIntegral w16)      -- TODO: fix
+
+
+-- | We don`t fitler in memory, so totalEntries is unknown, unless TxMeta Database counts them for us.
+-- It is possible due to some error, to have length ls < Page.
```

Comment:
Not sure  understand this comment "It is possible due to some error, to have length ls < Page.". Can you explain/reword?

Reply:
@edsko Assume we find a Tx in the TxMeta table, but then for some reason, we don`t find any inputs in the InputsTable. The current implementation would exclude such a Tx (see TxMeta.toValidKernelTxMeta the Nothing case). In this case the returned page would have less elements than it should, even though there may be sufficient total entries.

Reply:
Ok, perhaps just change the comment to spell this out in a bit more detail. It's not that important to fix it since transactions with no inputs should not exist, but it's good to document it.